File: scripts/exportar-bandeiras.js
#!/usr/bin/env node
// scripts/exportar-bandeiras.js
//
// Equivalente Node/sharp de scripts/exportar-bandeiras.py — exporta as 20 bandeiras de
// imagens/bandeiras/<ISO2>.svg para <ISO2>.webp 600x400 (3:2, ver DECISÃO 3:2 abaixo).
//
// POR QUE EXISTE UM SIBLING .js (cross-lane, frente de design, 2026-07-26): a tarefa pediu pra
// ESTENDER scripts/exportar-bandeiras.py pro novo alvo 3:2. A lógica de crop foi de fato
// atualizada lá (é a versão canônica/legível da decisão). Mas rodar esse .py NESTA máquina
// esbarrou num bloqueio de ambiente real, não um bug do script: cairosvg precisa da biblioteca
// nativa libcairo (cairo-2.dll), que no Windows só vem via um instalador separado (GTK3
// runtime) — não é algo que `pip install` resolve sozinho, e instalar um runtime de sistema
// inteiro só pra rodar 1 script parecia desproporcional (e é uma mudança de sistema que não é
// claramente reversível). `sharp` (Node) faz o mesmo trabalho (rasteriza SVG via libvips, que
// JÁ vem embutido no binário pré-compilado do pacote — sem instalador de sistema separado) e
// este projeto já roda Node pra tudo mais. Este arquivo é o que RODOU de fato nesta tarefa; o
// .py continua a documentação/lógica de referência (portável pra qualquer ambiente com cairo
// já disponível) — os dois devem ficar sincronizados se o alvo mudar de novo.
//
// DECISÃO 3:2 (correção de causa raiz, revisão do dono 2026-07-26): o acervo 1:1 600x600
// (commit anterior) forçava TODA bandeira a um corte quadrado, cortando o design de propósito
// mesmo nas 14 (de 20) bandeiras cuja proporção OFICIAL já é exatamente 3:2 — corte
// desnecessário nessas, e ainda mais corte que o necessário nas outras 6. 600x400 (3:2) casa
// com a proporção mais comum do acervo real (medido: AT/CN/ES/FR/GR/IN/IT/JP/KR/LB/MA/PE/PT/TH
// = 14 bandeiras EXATAMENTE 1.5 de razão nativa — corte ZERO nessas). As 6 restantes, medidas
// (não assumidas) direto do SVG:
//   MAIS LARGA que 3:2 (corta LARGURA): DE 1.6667 (5:3) | MX 1.75 (7:4) | US 1.9 (19:10) |
//     HU 2.0 (2:1, a mais larga do acervo)
//   MAIS ESTREITA que 3:2 (corta ALTURA): BR 1.4286 (10:7) | DK 1.3214 (~28:37)
//
// A EXCEÇÃO DOS EUA — NÃO REMOVER (mesma razão do .py original, ver lá o detalhe completo): o
// cantão fica no canto SUPERIOR ESQUERDO; um corte centralizado da largura o cortaria pela
// metade. ANCORA['US'] = 'left' ancora o corte na borda esquerda, preservando o cantão inteiro
// e cortando só do lado das listras (direita). Nenhuma outra bandeira do acervo tem emblema
// fora do centro que precise do mesmo tratamento (BR e MX têm emblema centralizado
// horizontalmente; DK tem a cruz nórdica com a barra horizontal no centro vertical — os cortes
// de largura/altura das outras 5 são centralizados, `position: 'centre'`).
//
// ZOOM MÍNIMO (item 2 da rodada, regra nova em docs/DESIGN-TOKENS.md): a área de mídia do tile
// agora casa a proporção do asset (3:2), então o `fit: 'cover'` do sharp já é o corte mínimo
// que cobre a caixa — não precisa de scale(1.15) extra no CSS pra esconder bordas de blur forte
// (blur ficou leve nesta rodada, ver --flag-blur/--flag-mosaic-blur em css/style.css).
//
// DEPENDÊNCIA: npm install sharp (só esta, sem system installer separado — ver nota acima).
//
// COMO RODAR (da raiz do repo)
//   node scripts/exportar-bandeiras.js

const fs = require("fs");
const path = require("path");
const sharp = require("sharp");

const DIR = path.join(__dirname, "..", "imagens", "bandeiras");
const TARGET_W = 600;
const TARGET_H = 400;
const TARGET_RATIO = TARGET_W / TARGET_H; // 1.5

// 'left' = ancora o corte na borda esquerda (preserva cantão) | 'centre' = default
const ANCORA = { US: "left" };

async function exportarUma(iso) {
  const svgPath = path.join(DIR, iso + ".svg");
  const nativa = await sharp(svgPath).metadata();
  const nativeRatio = nativa.width / nativa.height;

  // density fixa (ex.: 600) quebra pros 2 lados: SVG com viewBox pequeno (AT é só "9x6" de
  // unidades) rasteriza borrado/blocado; SVG com viewBox já grande (US é "7410x3900") explode
  // pro limite de pixels do sharp. Calibrado POR ARQUIVO a partir do próprio viewBox nativo,
  // mirando ~1600px no lado maior antes do corte/resize final — mesma ideia do "renderiza a 2x
  // o alvo, evita serrilhado" do .py original, só que a base de cálculo agora é o viewBox real
  // de cada SVG, não um valor de densidade fixo que ignora a escala interna de cada arquivo.
  const ladoMaiorNativo = Math.max(nativa.width, nativa.height);
  const density = 72 * (1600 / ladoMaiorNativo);

  // Renderiza no tamanho calibrado acima e já corta pro 3:2 num só passo via fit:'cover'.
  const buf2x = await sharp(svgPath, { density })
    .resize(TARGET_W * 2, TARGET_H * 2, { fit: "cover", position: ANCORA[iso] || "centre" })
    .flatten({ background: "#ffffff" })
    .toBuffer();

  const final = sharp(buf2x).resize(TARGET_W, TARGET_H, { kernel: sharp.kernel.lanczos3 });

  const [lossy, lossless] = await Promise.all([
    final.clone().webp({ quality: 82 }).toBuffer(),
    final.clone().webp({ lossless: true }).toBuffer(),
  ]);
  const usarLossless = lossless.length <= lossy.length;
  const escolhido = usarLossless ? lossless : lossy;

  const destino = path.join(DIR, iso + ".webp");
  fs.writeFileSync(destino, escolhido);

  return {
    iso,
    kb: escolhido.length / 1024,
    modo: usarLossless ? "lossless" : "lossy",
    nativeRatio,
    ancora: ANCORA[iso] || "centre",
    eixoCorte: nativeRatio > TARGET_RATIO ? "largura" : nativeRatio < TARGET_RATIO ? "altura" : "nenhum (já é 3:2)",
  };
}

async function main() {
  const svgs = fs
    .readdirSync(DIR)
    .filter((f) => f.endsWith(".svg"))
    .map((f) => f.slice(0, -4))
    .sort();
  if (!svgs.length) {
    console.error(`nenhum .svg em ${DIR}`);
    process.exit(1);
  }
  let total = 0;
  for (const iso of svgs) {
    const r = await exportarUma(iso);
    const marca = r.ancora !== "centre" ? "  <- ANCORA ESPECIAL (" + r.ancora + ")" : "";
    console.log(
      `  ${r.iso}.webp  ${r.kb.toFixed(1).padStart(6)} KB  ${r.modo.padEnd(9)} nativa ${r.nativeRatio.toFixed(4)}  corta ${r.eixoCorte}${marca}`
    );
    total += r.kb;
  }
  console.log(`\n${svgs.length} bandeiras, ${total.toFixed(0)} KB no total, 600x400 (3:2).`);
}

main().catch((e) => {
  console.error("falhou:", e.message);
  process.exit(1);
});

File: scripts/exportar-bandeiras.py
#!/usr/bin/env python3
"""
scripts/exportar-bandeiras.py

Exporta as 20 bandeiras de imagens/bandeiras/<ISO2>.svg para <ISO2>.webp 600x600.

POR QUE RASTER E NÃO SVG (decisão da frente de design, 26/07/2026)
  O app serve WEBP, não SVG, e isso resolve dois problemas de uma vez:
  1. PESO / PARSING. Bandeira com brasão é centenas de nós de path: MX tinha 156 KB e PE 114 KB
     de SVG, que um celular modesto precisa parsear a cada render. Em webp 600px viram 16,6 KB e
     16,0 KB de bitmap. O acervo inteiro caiu de 349 KB para 138 KB.
     Exceção honesta: PT ficou MAIOR como webp (23,4 KB contra 11,3 KB) — a esfera armilar é
     detalhe fino que raster não comprime bem. Não muda a decisão; fica registrado.
  2. PROPORÇÃO. Cada bandeira tem sua proporção oficial e elas NÃO são uniformes: AT e TH são 3:2,
     Brasil 10:7, México 7:4, EUA 19:10. Num slot de tamanho fixo isso daria larguras diferentes
     ou letterbox, e exigiria caso especial no CSS. Cortando 1:1 no export, a uniformidade é
     garantida por construção e o CSS trata bandeira igual a imagem de categoria — mesmo contrato.

  Os SVG originais continuam no repo como INSUMO (ver .gitignore): são a única fonte para
  reexportar noutro tamanho, e a fonte externa pode sumir.

A EXCEÇÃO DOS EUA — NÃO REMOVER
  Crop 1:1 centralizado funciona para 19 das 20. Para os EUA, não: o cantão (as 50 estrelas) fica
  no canto SUPERIOR ESQUERDO, e um corte central o deixaria de fora — a bandeira viraria listras
  vermelhas e brancas, irreconhecível. Por isso ANCORA_X['US'] = 0.0, que ancora o corte na borda
  esquerda e preserva o cantão inteiro. Qualquer bandeira nova com emblema fora do centro precisa
  da mesma análise.

LOSSY OU LOSSLESS
  Decidido POR ARQUIVO, não por regra: gera os dois e fica o menor. Bandeira de faixa chapada
  (AT, DE, FR, IT, TH, HU) comprime melhor lossless — 0,2 KB. Bandeira com brasão comprime melhor
  lossy. Fixar um dos dois pro acervo todo desperdiçaria de um lado ou do outro.

DEPENDÊNCIAS (as únicas do projeto fora de Node — de propósito, é build step, não runtime)
  pip install cairosvg pillow

COMO RODAR (da raiz do repo)
  python3 scripts/exportar-bandeiras.py
"""

import glob
import io
import os
import sys

try:
    import cairosvg
    from PIL import Image
except ImportError:
    sys.exit("faltam dependências: pip install cairosvg pillow")

DIR = os.path.join(os.path.dirname(__file__), "..", "imagens", "bandeiras")
LADO = 600

# 0.0 = ancora o corte na borda esquerda | 0.5 = centraliza (default)
ANCORA_X = {"US": 0.0}   # ver "A EXCEÇÃO DOS EUA" no topo


def exportar(svg):
    iso = os.path.basename(svg)[:-4]
    # renderiza a 2x o lado final: dá margem de reamostragem e evita serrilhado na diagonal
    png = cairosvg.svg2png(url=svg, output_height=LADO * 2)
    im = Image.open(io.BytesIO(png)).convert("RGBA")
    im = Image.alpha_composite(Image.new("RGBA", im.size, (255, 255, 255, 255)), im).convert("RGB")

    w, h = im.size
    if w < h:                                    # bandeira mais alta que larga: escala pela largura
        im = im.resize((LADO * 2, int(h * LADO * 2 / w)), Image.LANCZOS)
        w, h = im.size

    lado = min(w, h)
    x0 = int((w - lado) * ANCORA_X.get(iso, 0.5))
    y0 = (h - lado) // 2
    quad = im.crop((x0, y0, x0 + lado, y0 + lado)).resize((LADO, LADO), Image.LANCZOS)

    a, b = io.BytesIO(), io.BytesIO()
    quad.save(a, "WEBP", quality=82, method=6)
    quad.save(b, "WEBP", lossless=True, method=6)
    sem_perda = b.tell() <= a.tell()

    destino = os.path.join(DIR, iso + ".webp")
    quad.save(destino, "WEBP", **({"lossless": True, "method": 6} if sem_perda else {"quality": 82, "method": 6}))
    return iso, os.path.getsize(destino) / 1024, "lossless" if sem_perda else "lossy", ANCORA_X.get(iso, 0.5)


def main():
    svgs = sorted(glob.glob(os.path.join(DIR, "*.svg")))
    if not svgs:
        sys.exit(f"nenhum .svg em {DIR}")
    total = 0.0
    for svg in svgs:
        iso, kb, modo, ax = exportar(svg)
        marca = "  <- ANCORA ESPECIAL" if ax != 0.5 else ""
        print(f"  {iso}.webp  {kb:6.1f} KB  {modo}{marca}")
        total += kb
    print(f"\n{len(svgs)} bandeiras, {total:.0f} KB no total.")


if __name__ == "__main__":
    main()
